src/web-worker/amp-worker.js

```diff
@@ -0,0 +1,174 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {calculateEntryPointScriptUrl} from '../service/extension-location';
+import {dev} from '../log';
+import {fromClass} from '../service';
+import {isExperimentOn} from '../experiments';
+import {getMode} from '../mode';
+
+const TAG = 'web-worker';
+
+/**
+ * @typedef {{
+ *   method: string,
+ *   args: !Array,
+ *   id: number,
+ * }}
+ */
+export let ToWorkerMessageDef;
+
+/**
+ * @typedef {{
+ *   method: string,
+ *   returnValue: *,
+ *   id: number,
+ * }}
+ */
+export let FromWorkerMessageDef;
+
+/**
+ * Invokes function named `method` with args `opt_args` on the web worker
+ * and returns a Promise that will be resolved with the function's return value.
+ * @note Currently only works in a single entry point.
+ * @param {!Window} win
+ * @param {string} method
+ * @param {!Array=} opt_args
+ * @return {!Promise}
+ */
+export function invokeWebWorker(win, method, opt_args) {
+  if (!isExperimentOn(win, TAG)) {
+    return Promise.reject(`Experiment "${TAG}" is disabled.`);
+  }
+  if (!win.Worker) {
+    return Promise.reject('Worker not supported in window: ' + win);
+  }
+  const worker = fromClass(win, 'amp-worker', AmpWorker);
+  return worker.sendMessage_(method, opt_args || []);
+}
+
+/**
+ * @param {!Window} win
+ * @return {!AmpWorker}
+ * @visibleForTesting
+ */
+export function ampWorkerForTesting(win) {
+  return fromClass(win, 'amp-worker', AmpWorker);
+}
+
+/**
+ * A Promise-based API wrapper around a single Web Worker.
+ * @private
+ */
+class AmpWorker {
+  /**
+   * @param {!Window} win
+   */
+  constructor(win) {
+    /** @const @private {!Window} */
+    this.win_ = win;
+
+    const url =
+        calculateEntryPointScriptUrl(location, 'ww', getMode().localDev);
+    /** @const @private {!Worker} */
+    this.worker_ = new win.Worker(url);
+    this.worker_.onmessage = this.receiveMessage_.bind(this);
+
+    /**
+     * Maps method names to the promise executors for in-flight invocations of
+     * those methods. E.g. messages['foo'][3] contains the {resolve, reject}
+     * functions for the third concurrent invocation of 'foo'.
+     *
+     * @const @private {
+     *   !Object<string,
+     *     !Array<({resolve: !Function, reject: !Function}|undefined)>
+     *   >
+     * }
+     */
+    this.messages_ = Object.create(null);
+  }
+
+  /**
+   * Sends a method invocation request to the worker and returns a Promise.
+   * @param {string} method
+   * @param {!Array} args
+   * @return {!Promise}
+   * @private
+   */
+  sendMessage_(method, args) {
+    const promise = new Promise((resolve, reject) => {
+      if (!this.messages_[method]) {
```

Comment:
This works, but it might be simpler to have a map of outstanding messages with a simple running index that you increment for each message.